File: solutions/python/rail-fence-cipher/1/rail_fence_cipher.py
""" Module docstring """
def encode(message, rails):
    """ function docstring """
    rail = []
    for ri in range(rails):
        rail.append(" " * len(message))
    ri = 0
    dr = 1
    for ti in range(len(message)):
        rail[ri] = rail[ri][:ti] + message[ti] + rail[ri][ti+1:]
        ri += dr
        if ri == rails:
            dr = -1
            ri = rails - 2
        elif ri == -1:
            dr = 1
            ri = 1
    ec = ''
    for ri in range(rails):
        for ti in range(len(message)):
            if rail[ri][ti] != ' ':
                ec += rail[ri][ti]
    return ec


def decode(encoded_message, rails):
    """ function docstring """
    rail = []
    for ri in range(rails):
        rail.append("-" * (len(encoded_message)+5))
    dc_len = len(encoded_message)
    sl=dc_len//( 2 * rails - 2)
    si = []
    nc = []

    for ri in range(rails):
        if ri == 0:
            si.append(0)
            nc.append(sl)
            if dc_len % rails != 0:
                nc[0] += 1
        elif ri < rails - 1:
            si.append(sum(nc))
            nc.append(2*sl)
        else:
            si.append(sum(nc))
            nc.append(sl)
    if sum(nc) < dc_len:
        si[2] += 1
    dc = ''
    ri = 0
    dr = 1
    for ti in range(len(encoded_message)):
        rail[ri]= rail[ri][:ti] + encoded_message[si[ri]] + rail[ri][ti+1:]
        dc += rail[ri][ti]
        si[ri] += 1
        ri = ri + dr
        if ri >= rails:
            dr = -1
            ri = rails - 2
        elif ri < 0:
            dr = 1
            ri = 1

    return dc

File: solutions/c/rail-fence-cipher/1/rail_fence_cipher.c
#include "rail_fence_cipher.h"
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

char *encode(char *text, size_t rails){
    size_t tl=strlen(text);
    char ** rail = malloc(rails*sizeof(char *));
    for (size_t ri=0; ri<rails;ri++){
        rail[ri]=malloc((tl+1)*sizeof(char));
        size_t rli;
        for (rli=0;rli<tl;rli++){
            rail[ri][rli]=' ';
        }
        rail[ri][rli]='\0';
    }
    size_t ri=0;
    size_t rli=0;
    size_t dir=1;
    for (size_t ti=0;ti<tl;ti++){
        if(text[ti]!=' ') rail[ri][rli++]=text[ti];
        ri+=dir;
        if (ri==rails-1||ri==0) {
            dir=-1*dir;
        }
    }

    char * ec=malloc((tl+1)*sizeof(char));
    size_t eci = 0;
    for (size_t ri=0; ri<rails; ri++) {
        for (size_t rli=0; rli<tl; rli++) {
            if (rail[ri][rli]!=' ') ec[eci++] = rail[ri][rli];
        }
    }
    ec[eci]='\0';
    return ec;
}

char *decode(char *ciphertext, size_t rails){
    size_t cdl=strlen(ciphertext);
    char ** rail = malloc(rails*sizeof(char *));
    char * text=malloc((cdl+1)*sizeof(char));
    for (size_t ri=0; ri<rails;ri++){
        rail[ri]=malloc((cdl+1)*sizeof(char));
        size_t rli;
        for (rli=0;rli<cdl;rli++){
            rail[ri][rli]=' ';
        }
        rail[ri][rli]='\0';
    }

    int * r = calloc(rails,sizeof(int));
    int * nc = calloc(rails,sizeof(int));
    int * sti= calloc(rails,sizeof(int));

    nc[0] = cdl / (2 * rails - 2);
    if (cdl % (2 * rails - 2)!=0) nc[0]++;
    sti[0] = 0;

    for (int ri=1; ri < (int)(rails-1); ri++){
        nc[ri] =  cdl / (rails-1); 
        sti[ri] = sti[ri-1] + nc[ri-1];
    }

    nc[rails-1] = cdl / (2 * rails - 2);
    sti[rails-1] = nc[0]+(rails-2)*nc[1];       

    int rli=0;
    int ri=0;
    int dir = 1;
    for(rli=0;rli<(int)cdl;rli++){

        rail[ri][rli]=ciphertext[sti[ri]+r[ri]];
        r[ri]++;

        ri+=dir;
        if (ri==(int)rails) {
            dir=-1;
            ri=rails-2;
        } else if (ri==0) {
            dir=1;
            ri=0;
        }
    
    }

    for (size_t ri = 0; ri<rails; ri++) {
        rail[ri][rli]='\0';
    }
    int ti;
    rli = 0;
    ri = 0;
    dir = 1;
    for (ti=0;ti<(int)cdl;ti++){
        text[ti]=rail[ri][rli++];
        ri+=dir;
        if (ri==(int)rails-1||ri==0) {
            dir=-1*dir;
        }
    }
    text[ti]='\0';
    return text;

}
